File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/MiddlewareException.cs
using System;

namespace Iaiso.Middleware;

/// <summary>Base type for middleware-induced failures.</summary>
public abstract class MiddlewareException : Exception
{
    protected MiddlewareException(string message) : base(message) {}
    protected MiddlewareException(string message, Exception inner) : base(message, inner) {}

    /// <summary>Raised when <c>RaiseOnEscalation</c> is true and the engine has escalated.</summary>
    public sealed class EscalationRaised : MiddlewareException
    {
        public EscalationRaised() : base("execution escalated; raise-on-escalation enabled") {}
    }

    /// <summary>Raised when the execution is locked.</summary>
    public sealed class Locked : MiddlewareException
    {
        public Locked() : base("execution locked") {}
    }

    /// <summary>Raised when the upstream provider call failed.</summary>
    public sealed class Provider : MiddlewareException
    {
        public Provider(string message) : base("provider error: " + message) {}
        public Provider(string message, Exception inner)
            : base("provider error: " + message, inner) {}
    }
}

File: IAIso-v5.0/core/iaiso-java/iaiso-middleware/src/main/java/io/iaiso/middleware/MiddlewareException.java
package io.iaiso.middleware;

/** Base type for middleware-induced failures. */
public abstract class MiddlewareException extends RuntimeException {
    private static final long serialVersionUID = 1L;
    protected MiddlewareException(String message) { super(message); }
    protected MiddlewareException(String message, Throwable cause) { super(message, cause); }

    /** Raised when {@code raiseOnEscalation} is true and the engine has escalated. */
    public static final class EscalationRaised extends MiddlewareException {
        private static final long serialVersionUID = 1L;
        public EscalationRaised() { super("execution escalated; raise-on-escalation enabled"); }
    }

    /** Raised when the execution is locked. */
    public static final class Locked extends MiddlewareException {
        private static final long serialVersionUID = 1L;
        public Locked() { super("execution locked"); }
    }

    /** Raised when the upstream provider call failed. */
    public static final class Provider extends MiddlewareException {
        private static final long serialVersionUID = 1L;
        public Provider(String message) { super("provider error: " + message); }
        public Provider(String message, Throwable cause) {
            super("provider error: " + message, cause);
        }
    }
}
